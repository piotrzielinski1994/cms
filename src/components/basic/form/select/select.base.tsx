import { EnhancedHtmlProps, HtmlProps } from '@/utils/html/html.types';
import { forwardRef } from 'react';
import Form from '../root/form';

// prettier-ignore
type SelectProps = EnhancedHtmlProps<'select', {
  name: string;
  value?: string;
}>;

const Wrapper = (props: HtmlProps['div']) => {
  return <div {...props} />;
};

const Native = forwardRef<HTMLSelectElement, Omit<SelectProps, 'placeholder'>>((props, ref) => {
  return <select ref={ref} {...props} />;
});

const Option = forwardRef<HTMLOptionElement, HtmlProps['option']>((props, ref) => {
  return <option ref={ref} {...props} />;
});

const SelectBase = {
  Root: Form.Group,
  Label: Form.Label,
  Wrapper,
  Native,
  Option,
  Error: Form.Error,
};

export default SelectBase;
