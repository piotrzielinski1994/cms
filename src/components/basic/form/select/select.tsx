import { cn } from '@/utils/tailwind';
import { ChevronDown } from 'lucide-react';
import { DetailedHTMLProps, SelectHTMLAttributes } from 'react';
import { Control, FieldValues, Path, useController } from 'react-hook-form';
import Form from '../root/form';
import { inputClassNames } from '../text-input/text-input';

type SelectProps = Omit<
  DetailedHTMLProps<SelectHTMLAttributes<HTMLSelectElement>, HTMLSelectElement>,
  'name'
> & {
  name: string;
  error?: string;
  placeholder?: string;
  options: {
    value: string;
    label: string;
  }[];
};

type SelectContainerProps<T extends FieldValues> = Omit<SelectProps, 'name'> & {
  control: Control<T>;
  name: Path<T>;
};

const Select = ({ error, options, placeholder = '', className, ...props }: SelectProps) => {
  return (
    <div className="flex flex-col gap-2">
      <div className="grid items-center">
        <select
          defaultValue=""
          className={cn(
            'appearance-none',
            { 'cursor-pointer': !props.disabled },
            inputClassNames.input({ isValid: !error }),
            'col-start-1 row-start-1',
            className,
          )}
          {...props}
        >
          <option value="" disabled className="hidden">
            {placeholder}
          </option>
          {options.map((o) => {
            return (
              <option key={o.value} value={o.value}>
                {o.label}
              </option>
            );
          })}
        </select>
        <ChevronDown size="1rem" className="col-start-1 row-start-1 justify-self-end mr-1" />
      </div>
      <Form.Error>{error}</Form.Error>
    </div>
  );
};

const SelectContainer = <T extends FieldValues>(props: SelectContainerProps<T>) => {
  const { control, name, ...rest } = props;
  const { field, fieldState } = useController({ control, name });
  return (
    <Select
      error={fieldState.error?.message}
      {...rest}
      {...field}
      onChange={(e) => {
        const value = e.target.value === '' ? null : e.target.value;
        field.onChange(value);
      }}
    />
  );
};

export { Select, SelectContainer };
