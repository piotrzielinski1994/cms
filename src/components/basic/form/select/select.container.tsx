import { ComponentProps } from 'react';
import { Control, FieldValues, Path, useController } from 'react-hook-form';
import { Select } from './select';

type SelectContainerProps<T extends FieldValues> = {
  control: Control<T>;
  name: Path<T>;
} & Omit<ComponentProps<typeof Select>, 'name'>;

const SelectContainer = <T extends FieldValues>(props: SelectContainerProps<T>) => {
  const { control, name, ...rest } = props;
  const { field, fieldState } = useController({ control, name });
  return (
    <Select
      error={fieldState.error?.message}
      {...rest}
      {...field}
      onChange={(e) => {
        const value = e.target.value === '' ? null : e.target.value;
        field.onChange(value);
      }}
    />
  );
};

export { SelectContainer };
