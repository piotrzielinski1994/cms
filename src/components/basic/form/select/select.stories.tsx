import Form from '@/components/basic/form/root/form';
import type { Meta, StoryObj } from '@storybook/react';
import { useTranslations } from 'next-intl';
import { useState, type ComponentProps } from 'react';
import { Select as SelectComponent } from './select';

type Args = ComponentProps<typeof SelectComponent> & {
  label?: string;
};

const ControlledInput = (props: Args) => {
  const [value, setValue] = useState('');
  return (
    <SelectComponent {...props} onChange={(e) => setValue(e.currentTarget.value)} value={value} />
  );
};

const meta: Meta<ComponentProps<typeof ControlledInput>> = {
  component: ControlledInput,
  title: 'Components/Basic/Form/Select',
  argTypes: {
    label: { control: 'text' },
    placeholder: { control: 'text' },
    error: { control: 'text' },
    options: { control: 'object' },
    disabled: { control: 'boolean' },
  },
  args: {
    label: 'Select Input',
    placeholder: 'Choose value',
    options: [
      { value: 'one', label: 'One' },
      { value: 'two', label: 'Two' },
    ],
    disabled: false,
  },
};

const Render = ({ label, ...args }: Args) => {
  const t = useTranslations('storybook.basic.form.textInput');

  return (
    <>
      <Form.Group>
        <Form.Label htmlFor="input1">{label}</Form.Label>
        <ControlledInput {...args} id="input1" name="input1" />
      </Form.Group>
      <Form.Group>
        <Form.Label htmlFor="input2">{t('disabled.label')}</Form.Label>
        <SelectComponent
          id="input2"
          name="input2"
          placeholder={t('disabled.placeholder')}
          disabled
          options={[]}
        />
      </Form.Group>
      <Form.Group>
        <Form.Label htmlFor="input3">{t('invalid.label')}</Form.Label>
        <SelectComponent
          {...args}
          id="input3"
          name="input3"
          value={t('invalid.value')}
          error={t('invalid.error')}
          options={args.options}
        />
      </Form.Group>
    </>
  );
};

const Select: StoryObj<typeof SelectComponent> = { render: Render };

export { Select };
export default meta;
