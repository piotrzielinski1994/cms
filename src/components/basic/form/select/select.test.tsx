import { render } from '@testing-library/react';
import { ComponentProps } from 'react';
import { describe, expect, it } from 'vitest';
import { axe } from 'vitest-axe';
import Form from '../root/form';
import { Select } from './select';

describe('Select', () => {
  const defaultProps = {
    name: 'name',
    id: 'id',
    options: [1, 2, 3].map((it) => ({
      value: String(it),
      label: `Option ${it}`,
    })),
  } satisfies ComponentProps<typeof Select>;

  it('should have no accessibility violations', async () => {
    const { container } = render(
      <Form.Group>
        <Form.Label htmlFor={defaultProps.id}>Label</Form.Label>
        <Select {...defaultProps} />
      </Form.Group>,
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', async () => {
    const { container } = render(<Select {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
