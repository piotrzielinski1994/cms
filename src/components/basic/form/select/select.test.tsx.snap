// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select > should match the snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <div
      class="grid items-center"
    >
      <select
        class="appearance-none cursor-pointer p-2 border border-solid border-current bg-input placeholder-foreground/50 [&:not(:disabled)]:hover:border-foreground/90 [&:not(:disabled)]:hover:ring-foreground/90 disabled:text-foreground/50 disabled:placeholder-foreground/50 outline-none ring-inset focus-within:ring-1 ring-current col-start-1 row-start-1"
        id="id"
        name="name"
      >
        <option
          class="hidden"
          disabled=""
          selected=""
          value=""
        />
        <option
          value="1"
        >
          Option 1
        </option>
        <option
          value="2"
        >
          Option 2
        </option>
        <option
          value="3"
        >
          Option 3
        </option>
      </select>
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-down col-start-1 row-start-1 justify-self-end mr-1"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m6 9 6 6 6-6"
        />
      </svg>
    </div>
    <span
      class="min-h-[1em] text-sm text-red-500 leading-none"
      role="alert"
    />
  </div>
</div>
`;
